### PR TITLE
Initial writer support

### LINKED ISSUE #1: writer: encoding of FIT files

### PR DESCRIPTION
**Pull request by usedbytes**
Implements support for encode, with some known limitations (e.g. #36, #37)

Fixes #1

### CODE REVIEW

**Comment by usedbytes** [comment]
The first commit is necessary for me, otherwise fitgen segfaults. I'm not sure why I'm seeing it if you aren't, but I think it should be harmless.

File: header.go [hunk 1]
```diff
@@ -152,3 +153,44 @@ func (h Header) String() string {
...
+	}
+
+	err = binary.Write(buf, binary.LittleEndian, h.DataType)
+	if err != nil {
+		return nil, err
+	}
+
+	h.CRC = dyncrc16.Checksum(buf.Bytes())
+
+	if h.Size == 14 {
```
Comment (tormoder) [inline header.go]: Can we use `headerSizeCRC` from `consts.go`?

File: writer.go [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+}
+
+type encodeMesgDef struct {
+	globalMesgNum MesgNum
+	localMesgNum  byte
+	fields        []*field
+}
+
+func (e *encoder) writeMesg(mesg reflect.Value, def *encodeMesgDef) error {
+	hdr := def.localMesgNum & 0xF
```
Comment (tormoder) [inline writer.go]: Can we use `localMesgNumMask` from `consts.go`?

File: writer.go [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+func Encode(w io.Writer, file *File, arch binary.ByteOrder) error {
+	buf := &bytes.Buffer{}
+	enc := &encoder{
+		w:    buf,
+		arch: arch,
+	}
+
+	// XXX: Is there a better way to do this with reflection?
+	var data reflect.Value
+	switch file.Type() {
```
Comment (tormoder) [inline writer.go]: Would it make sense to add a default case here with an error return just in case?

**Comment by usedbytes** [comment]
Thanks for the review! All the comments make sense, I'll send a respin with them fixed.

In the process of implementing your suggestions, though, I've found that the DecodeEncodeDecode test wasn't really doing anything after a last minute change, because I `return` instead of `continue` on skipped files. Now I'm seeing some bugs related to Array handling - so I'll fix that up first.

I'm about to go away for a couple of weeks, but I will try to get the update pushed while I'm travelling.

File: header.go (hunk 1 above)
Comment (usedbytes) [reply header.go]: Done

File: writer.go (hunk 2 above)
Comment (usedbytes) [reply writer.go]: Done, found another instance and opportunity for mesgDefinitionMask too.

File: writer.go (hunk 3 above)
Comment (usedbytes) [reply writer.go]: Done

**Comment by usedbytes** [comment]
So it wasn't as bad as I thought. I had to "blacklist" a couple more files in the round-trip test, but there's still 12 files which get run through it. I opened #36 and #37 for the major limitations, and tagged the `skipEncode`s with those issue numbers where appropriate.

I added some text to the merge message at the top, I hope that works. I've not done many Github pull requests before :smile:

**Comment by tormoder** [comment]
Great, thank you for writing the down the limitations. 

I'll merge this and tag a 0.4.0 release soon.

Thanks again!

**Pull request merged**